### PR TITLE
Code cleanup, which may fix Faces TCK WebSocket test

### PR DESCRIPTION
**Pull request by dmatej**
- Disclaimer: I don't like it.
- Two commits: first is just readability, second fixes the issue. You can review just the second as the first is trivial (but larger).
- Javadocs were obsoleted and parents usually had better or same. In one was explicitly told, that addFilter can return null - TyrusServletContainerInitializer doesn't expect that. But that's not all - I don't see real reason why it could be null in this case. What I do see is that there's many synchronized blocks which cannot protect from race conditions. But it seems this is just a single thread ... 
- The original stacktrace:

```
[2023-01-29T09:34:56.177183+01:00] [GlassFish 7.0] [SEVERE] [AS-WEB-CORE-00108] [jakarta.enterprise.web.core] [tid: _ThreadID=68 _ThreadName=admin-listener(5)] [levelValue: 1000] [[
ContainerBase.addChild: start: 
org.apache.catalina.LifecycleException: org.apache.catalina.LifecycleException: Error invoking ServletContainerInitializer org.glassfish.tyrus.servlet.TyrusServletContainerInitializer
at org.apache.catalina.core.StandardContext.start(StandardContext.java:4469)
at com.sun.enterprise.web.WebModule.start(WebModule.java:552)
at org.apache.catalina.core.ContainerBase.addChildInternal(ContainerBase.java:935)
at org.apache.catalina.core.ContainerBase.addChild(ContainerBase.java:917)
at org.apache.catalina.core.StandardHost.addChild(StandardHost.java:644)
at com.sun.enterprise.web.WebContainer.loadWebModule(WebContainer.java:1795)
at com.sun.enterprise.web.WebContainer.loadWebModule(WebContainer.java:1483)
at com.sun.enterprise.web.WebApplication.start(WebApplication.java:88)
at org.glassfish.internal.data.EngineRef.start(EngineRef.java:97)
at org.glassfish.internal.data.ModuleInfo.start(ModuleInfo.java:262)
at org.glassfish.internal.data.ApplicationInfo.start(ApplicationInfo.java:353)
at com.sun.enterprise.v3.server.ApplicationLifecycle.deploy(ApplicationLifecycle.java:535)
at com.sun.enterprise.v3.server.ApplicationLifecycle.deploy(ApplicationLifecycle.java:259)
at org.glassfish.deployment.admin.DeployCommand.execute(DeployCommand.java:467)
at com.sun.enterprise.v3.admin.CommandRunnerImpl$2$1.run(CommandRunnerImpl.java:529)
at com.sun.enterprise.v3.admin.CommandRunnerImpl$2$1.run(CommandRunnerImpl.java:525)
at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
at java.base/javax.security.auth.Subject.doAs(Subject.java:376)
at com.sun.enterprise.v3.admin.CommandRunnerImpl$2.execute(CommandRunnerImpl.java:524)
at com.sun.enterprise.v3.admin.CommandRunnerImpl$3.run(CommandRunnerImpl.java:555)
at com.sun.enterprise.v3.admin.CommandRunnerImpl$3.run(CommandRunnerImpl.java:547)
at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
at java.base/javax.security.auth.Subject.doAs(Subject.java:376)
at com.sun.enterprise.v3.admin.CommandRunnerImpl.doCommand(CommandRunnerImpl.java:546)
at com.sun.enterprise.v3.admin.CommandRunnerImpl.doCommand(CommandRunnerImpl.java:1456)
at com.sun.enterprise.v3.admin.CommandRunnerImpl$ExecutionContext.execute(CommandRunnerImpl.java:1825)
at com.sun.enterprise.v3.admin.CommandRunnerImpl$ExecutionContext.execute(CommandRunnerImpl.java:1701)
at org.glassfish.admin.rest.utils.ResourceUtil.runCommand(ResourceUtil.java:221)
at org.glassfish.admin.rest.utils.ResourceUtil.runCommand(ResourceUtil.java:202)
at org.glassfish.admin.rest.utils.ResourceUtil.runCommand(ResourceUtil.java:241)
at org.glassfish.admin.rest.resources.TemplateListOfResource.createResource(TemplateListOfResource.java:305)
at org.glassfish.admin.rest.resources.TemplateListOfResource.post(TemplateListOfResource.java:142)
at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
at java.base/java.lang.reflect.Method.invoke(Method.java:568)
at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory.lambda$static$0(ResourceMethodInvocationHandlerFactory.java:52)
at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:134)
at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:177)
at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$ResponseOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:176)
at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:81)
at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:478)
at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:400)
at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:81)
at org.glassfish.jersey.server.ServerRuntime$1.run(ServerRuntime.java:261)
at org.glassfish.jersey.internal.Errors$1.call(Errors.java:248)
at org.glassfish.jersey.internal.Errors$1.call(Errors.java:244)
at org.glassfish.jersey.internal.Errors.process(Errors.java:292)
at org.glassfish.jersey.internal.Errors.process(Errors.java:274)
at org.glassfish.jersey.internal.Errors.process(Errors.java:244)
at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:265)
at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:240)
at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:697)
at org.glassfish.jersey.grizzly2.httpserver.GrizzlyHttpContainer.service(GrizzlyHttpContainer.java:367)
at org.glassfish.admin.rest.adapter.RestAdapter$2.service(RestAdapter.java:287)
at org.glassfish.admin.rest.adapter.RestAdapter.service(RestAdapter.java:156)
at com.sun.enterprise.v3.services.impl.ContainerMapper$HttpHandlerCallable.call(ContainerMapper.java:453)
at com.sun.enterprise.v3.services.impl.ContainerMapper.service(ContainerMapper.java:145)
at org.glassfish.grizzly.http.server.HttpHandler.runService(HttpHandler.java:174)
at org.glassfish.grizzly.http.server.HttpHandler.doHandle(HttpHandler.java:153)
at org.glassfish.grizzly.http.server.HttpServerFilter.handleRead(HttpServerFilter.java:196)
at org.glassfish.grizzly.filterchain.ExecutorResolver$9.execute(ExecutorResolver.java:88)
at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeFilter(DefaultFilterChain.java:246)
at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeChainPart(DefaultFilterChain.java:178)
at org.glassfish.grizzly.filterchain.DefaultFilterChain.execute(DefaultFilterChain.java:118)
at org.glassfish.grizzly.filterchain.DefaultFilterChain.process(DefaultFilterChain.java:96)
at org.glassfish.grizzly.ProcessorExecutor.execute(ProcessorExecutor.java:51)
at org.glassfish.grizzly.nio.transport.TCPNIOTransport.fireIOEvent(TCPNIOTransport.java:510)
at org.glassfish.grizzly.strategies.AbstractIOStrategy.fireIOEvent(AbstractIOStrategy.java:82)
at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.run0(WorkerThreadIOStrategy.java:83)
at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy$WorkerThreadRunnable.run(WorkerThreadIOStrategy.java:101)
at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.doWork(AbstractThreadPool.java:535)
at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.run(AbstractThreadPool.java:515)
at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: org.apache.catalina.LifecycleException: Error invoking ServletContainerInitializer org.glassfish.tyrus.servlet.TyrusServletContainerInitializer
at org.apache.catalina.core.StandardContext.callServletContainerInitializers(StandardContext.java:4546)
at com.sun.enterprise.web.WebModule.callServletContainerInitializers(WebModule.java:638)
at org.apache.catalina.core.StandardContext.start(StandardContext.java:4449)
... 73 more
Caused by: java.lang.NullPointerException: Cannot invoke "jakarta.servlet.FilterRegistration$Dynamic.setAsyncSupported(boolean)" because "reg" is null
at org.glassfish.tyrus.servlet.TyrusServletContainerInitializer.onStartup(TyrusServletContainerInitializer.java:111)
at org.apache.catalina.core.StandardContext.callServletContainerInitializers(StandardContext.java:4541)
... 75 more
]]
```

### CODE REVIEW

**Comment by dmatej** [comment]
This kind of synchronization doesn't prevent anything except CME. It creates an array of current Container values and these values can be changed in parallel. I did not see the thread which changes them in logs, but I am quite sure there is some.

```
    public Container[] findChildren() {
        synchronized (children) {
            return children.values().toArray(new Container[children.size()]);
        }
    }
```
And here is the code which can return null from addFilter. Eh, maybe it is not a race condition. It depends on the order of actions. Tree. I am quite sure we should revisit it later again ...
```
        if (host != null) {
            for (Container child : host.findChildren()) {
                if (child == this) {
                    // Our own context will be checked further down
                    continue;
                }
                if (((StandardContext) child).hasFilter(filter)) {
                    return null;
                }
            }
        }
```

File: appserver/web/web-core/src/main/java/org/apache/catalina/core/StandardContext.java [hunk 1]
```diff
@@ -741,28 +720,30 @@ public StandardContext() {
      */
     private List<String> orderedLibs;
 
-    // <jsp-config> related info aggregated from web.xml and web-fragment.xml
+    /** <jsp-config> related info aggregated from web.xml and web-fragment.xml */
```
Comment (pzygielo) [inline appserver/web/web-core/src/main/java/org/apache/catalina/core/StandardContext.java:723]: If this gets promoted to be javadoc, shouldn't `<` and `>` be escaped?

**Comment by dmatej** [comment]
mvn clean install -Pfastest -T4C && mvn clean install -Ptck -pl :glassfish-external-tck-faces

```
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  02:36 h
[INFO] Finished at: 2023-01-29T12:18:37+01:00
[INFO] ------------------------------------------------------------------------
```

**Comment by pzygielo** [comment]
> 
>     * Two commits: first is just readability, second fixes the issue. You can review just the second as the first is trivial (but larger).

I fail to see what is being fixed and where. With _- It is just a replacement of for i loops by enhanced loops._ - does it mean that just replacing one with another helps? Pointing to the lines would help me, for sure.

```
[2023-01-29T09:34:56.177183+01:00] [GlassFish 7.0] [SEVERE] [AS-WEB-CORE-00108] [jakarta.enterprise.web.core] [tid: _ThreadID=68 _ThreadName=admin-listener(5)] [levelValue: 1000] [[
ContainerBase.addChild: start: 
org.apache.catalina.LifecycleException: org.apache.catalina.LifecycleException: Error invoking ServletContainerInitializer org.glassfish.tyrus.servlet.TyrusServletContainerInitializer
  at org.apache.catalina.core.StandardContext.start(StandardContext.java:4469)
  at com.sun.enterprise.web.WebModule.start(WebModule.java:552)
```

What GF version was used for that (log says it's *GlassFish 7.0*, but from which commit it was built)?
StandardContext.java:4469 on current master (cd4346068) doesn't seem to be `start`, StandardContext.java:4546 doesn't seem to be `callServletContainerInitializers`.

**Comment by arjantijms** [comment]
I also don't see why this fixes anything, but the changes themselves are okay. It's just a mystery why this has any effect.

**Comment by dmatej** [comment]
The stack trace was made after first commit, the faces tck result after the second. And yes, for loop conversions are dangerous on code with bad design (I disabled automatic conversions in the second workspace, because it sometimes breaks things on GF; this time it fixed something, hurray :rofl:  ). 

I don't think we are done here, but I don't have another month to refactor whole catalina.

Similar issue was "fixed" last year by swallowing the exception. That was reverted in november if I remember well, since then were these tests passing.

I have a similar situation in OpenMQ start, there too I cannot find out why I am seeing NPE in logs. On several other places they disappeared along the path to GF 7.0.0 and 7.0.1, so it is improving, but ... more work required. Catalina does evil things, especially StandardContext, which is a "god object antipattern", calling other objects which call it back again.

**Comment by pzygielo** [comment]
https://github.com/eclipse-ee4j/glassfish/blob/63d2d7e17b91e7a8c44e21607c550cf55e5e69eb/api/src/main/java/jakarta/servlet/ServletContext.java#L811-L813

> or **null** if this ServletContext already contains a complete FilterRegistration for a filter with the given filterName or if the same filter instance has already been registered with this or another ServletContext in the same container


https://github.com/eclipse-ee4j/glassfish/blob/d1186aaca906e14c84994d44d68fe75e907bb65d/containers/servlet/src/main/java/org/glassfish/tyrus/servlet/TyrusServletContainerInitializer.java#L110-L111

**Comment by pzygielo** [comment]
I've got the exception as in opening post with the master as well.

Then I've resumed from `:websocket` and build completed successfully.

-----

Then I've built GF again from
```
$ git rev-parse @
a3f8b1b27a944256b9753eb6b1d87f53d3807e58
```

and ran `mvn clean install -Ptck -pl :glassfish-external-tck-faces` to get:

```
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  28:44 min
[INFO] Finished at: 2023-01-29T20:29:42+01:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-failsafe-plugin:3.0.0-M5:verify (default) on project websocket: There are test failures.
[ERROR] 
```
with this
```
[2023-01-29T20:29:40.230118+01:00] [GlassFish 7.0] [SEVERE] [AS-WEB-CORE-00108] [jakarta.enterprise.web.core] [tid: _ThreadID=46 _ThreadName=admin-listener(4)] [levelValue: 1000] [[
  ContainerBase.addChild: start: 
org.apache.catalina.LifecycleException: org.apache.catalina.LifecycleException: Error invoking ServletContainerInitializer org.glassfish.tyrus.servlet.TyrusServletContainerInitializer
	at org.apache.catalina.core.StandardContext.start(StandardContext.java:4469)
	at com.sun.enterprise.web.WebModule.start(WebModule.java:552)
	at org.apache.catalina.core.ContainerBase.addChildInternal(ContainerBase.java:935)
	at org.apache.catalina.core.ContainerBase.addChild(ContainerBase.java:917)
	at org.apache.catalina.core.StandardHost.addChild(StandardHost.java:644)
	at com.sun.enterprise.web.WebContainer.loadWebModule(WebContainer.java:1795)
	at com.sun.enterprise.web.WebContainer.loadWebModule(WebContainer.java:1483)
	at com.sun.enterprise.web.WebApplication.start(WebApplication.java:88)
	at org.glassfish.internal.data.EngineRef.start(EngineRef.java:97)
```
in server.log.

Then I've resumed from `:websocket` and build successfully went to next module (xhtmlMappingToFaceletByDefault).

Not sure it's the TCK doing something wrong or tyrus not handling `null` properly.
Not sure acb2a109ebbbbb46a9dea85c2f993f73fec0412b fixes that in GF.

**Comment by arjantijms** [comment]
Whether it's another coincidence as well, I started the Faces TCK adding the following to the runner pom:

```
<argument>-Dtck.mode=platform</argument>
```

And it seemed to reproduce more faithfully, but only tested a couple of times.

**Review by arjantijms — APPROVED** [review]

**Comment by arjantijms** [comment]
As the code cleanup has value by itself, I'll merge this to start with. We can/should examine the real issue still.

**Pull request merged**